tools/lint/src/main/kotlin/NonAndroidxNullabilityDetector.kt

```diff
@@ -0,0 +1,123 @@
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.LintFix
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import org.jetbrains.uast.UAnnotation
+import org.jetbrains.uast.UClass
+import org.jetbrains.uast.UDeclaration
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UField
+import org.jetbrains.uast.UMethod
+
+private val NULLABLE_ANNOTATIONS = listOf("Nullable", "CheckForNull")
+private val NOT_NULL_ANNOTATIONS = listOf("NonNull", "NotNull", "Nonnull")
+
+class NonAndroidxNullabilityDetector : Detector(), SourceCodeScanner {
+    companion object Issues {
+        private val IMPLEMENTATION = Implementation(
+                NonAndroidxNullabilityDetector::class.java,
+                Scope.JAVA_FILE_SCOPE
+        )
+
+        @JvmField
+        val NON_ANDROIDX_NULLABILITY = Issue.create(
+                id = "FirebaseNonAndroidxNullability",
+                briefDescription = "Use androidx nullability annotations.",
+
+                explanation = "Use androidx nullability annotations instead.",
+                category = Category.COMPLIANCE,
+                priority = 1,
+                severity = Severity.ERROR,
+                implementation = IMPLEMENTATION
+        )
+    }
+
+    override fun getApplicableUastTypes(): List<Class<out UElement>>? {
+        return listOf(UClass::class.java, UMethod::class.java, UField::class.java)
+    }
+
+    override fun createUastHandler(context: JavaContext): UElementHandler? {
+        return Visitor(context)
+    }
+
+    class Visitor(private val context: JavaContext) : UElementHandler() {
+        override fun visitClass(node: UClass) {
+            doVisit(node)
+        }
+
+        override fun visitMethod(node: UMethod) {
+            doVisit(node)
+            for (parameter in node.uastParameters) {
+                doVisit(parameter)
+            }
+        }
+
+        override fun visitField(node: UField) {
+            doVisit(node)
+        }
+
+        private fun doVisit(node: UDeclaration) {
+            for (annotation in node.annotations) {
+                if (isNullabilityAnnotation(annotation)) {
+                    ensureAndroidNullability(context, annotation)
+                }
+            }
+        }
+
+        private fun ensureAndroidNullability(context: JavaContext, annotation: UAnnotation) {
+            annotation.qualifiedName?.let { name ->
+                val path = name.split('.')
+                val packageName = path.subList(0, path.size - 1).joinToString(".")
+                if (packageName in listOf("androidx.annotation", "android.support.annotation")) {
```

Comment:
I actually prefer the style used in 

https://github.com/JetBrains/kotlin/blob/master/core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/JvmAnnotationNames.kt

where you declare a list of fully qualified nullable & nonnullable annotations. Then it's a bit clearer than checking for the packageName.

Reply:
Done